### ISSUE TITLE
Automatic snapshot naming

### ISSUE THREAD

**Issue by grantr** [elastic/elasticsearch#7939]
It would be useful if Elasticsearch could generate automatic snapshot names. This would be a POST to the repository instead of a PUT to a snapshot name.

```
curl -XPOST localhost:9200/_snapshot/my-repository
```

The name could be generated by incrementing a number (`1`, `2`, `3`, etc), getting the current timestamp, or generating a flake id.

To organize snapshots in a repository, the POST action could accept a prefix parameter.

```
curl -XPOST localhost:9200/_snapshot/my-repository?prefix=my-index -d '{"indices":"my-index"}'
```

**Comment by colings86** [comment]
I would be wary of just having an incrementing number as it doesn't containing any information about the point in time the snapshot relates to, but maybe we could use something like the combination of timestamp and flake ID so that the timestamp indicates the point in time the snapshot was created and the flake ID makes the name unique?

**Comment by grantr** [comment]
@colings86 the snapshot metadata already includes the timestamp, so it's not necessary for preserving snapshot creation time. Still, I support using the timestamp as id because timestamps are useful when looking at a list of snapshot names. Most of our snapshots are named with the pattern `prefix-2015-02-21_12-57-41` for natural ordering and so that it's easy to see when snapshots occurred.

**Comment by tlrx** [comment]
We looked at this today with @ywelsch and we came up with the idea of reusing the date math expression resolver that already exists for indices.

This way we could create a snapshot like
```
POST /_snapshot/my-repository/<snapshot-{now/d}>
```
and it would be resolved to `snapshot-2018.03.21`.
